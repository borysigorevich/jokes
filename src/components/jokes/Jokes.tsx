import React, {useEffect} from 'react'
import {Link, useParams} from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux';
import {CardGroup, ListGroup, Nav, Navbar, OverlayTrigger, Popover} from 'react-bootstrap';
import {
    getCategories,
    getJokes,
    getMoreJokes, JokeType,
    setInitialRender, setSortedJokes
} from '../../redux/store/joke-reducer';
import {RootState} from '../../redux/store/store';
import {CustomButton, CustomNavLink, FlexDiv, Styles} from '../../styled/JokesStyled';


const Jokes = () => {

    const jokesState = useSelector((state: RootState) => state.jokes)
    const dispatch = useDispatch()
    const params = useParams()

    useEffect(() => {
        if (jokesState.initialRender) {
            try {
                dispatch(getJokes(params.category ? params.category : 'all'))
                dispatch(setInitialRender())
            } catch (err) {
                console.log(err)
            }
        }
    }, [dispatch, jokesState.initialRender, params.category])

    useEffect(() => {
        try {
            params.category && dispatch(setSortedJokes(params.category))
        } catch (err) {
            console.log(err)
        }
    }, [dispatch, params.category])

    useEffect(() => {
        if (jokesState.initialRender) {
            try {
                dispatch(getCategories())
            } catch (err) {
                console.log(err)
            }
        }
    }, [dispatch, jokesState.initialRender])

    const handleClick = async (numOfJokes: string) => {
        try {
            dispatch(getMoreJokes(numOfJokes))
        } catch (err) {
            console.log(err)
        }
    }

    return (
        <Styles>
            <Navbar expand={'md'} className={'justify-content-between'}>
                <Navbar.Brand>Choose the category</Navbar.Brand>
                <Navbar.Toggle aria-controls={'navbar'}/>
                <Navbar.Collapse id={'navbar'} className={'justify-content-end'}>
                    <Nav>
                        <CustomNavLink as={Link}
                                       key={'all'}
                                       to={'/all'}
                                       className={'mb-2 mt-3 mt-sm-0 mb-md-0 me-md-2'}
                        >All</CustomNavLink>
                        {jokesState.categories.map((category: string, index) => <CustomNavLink as={Link}
                                                                                        key={category}
                                                                                        className={`mb-2 mb-md-0 me-md-${jokesState.categories.length - 1 === index ? 0 : 2}`}
                                                                                        to={'/' + category}>{category[0].toUpperCase() + category.substring(1)}</CustomNavLink>)}
                    </Nav>
                </Navbar.Collapse>
            </Navbar>
            <CardGroup className={'justify-content-center'}>
                <ListGroup as={'ul'} className={'w-100'}>
                    {jokesState.sortedJokes.length > 0
                        ? jokesState.sortedJokes.map((joke: JokeType) => (
                            <OverlayTrigger
                                trigger={'click'}
                                key={joke.id}
                                placement={'bottom'}
                                overlay={
                                    <Popover>
                                        <Popover.Body>
                                            {joke.joke}
                                        </Popover.Body>
                                    </Popover>
                                }>
                                <ListGroup.Item as={'li'} key={joke.id}
                                                className={'border border-secondary my-1 rounded'}>
                                    {joke.joke.length > 100 ? joke.joke.substring(0, 100) + '...' : joke.joke}
                                </ListGroup.Item>
                            </OverlayTrigger>))
                        : <p className={'fs-3 text-secondary'}>Ooops!! There are no jokes! Get Some!</p>
                    }
                </ListGroup>
            </CardGroup>
            <FlexDiv>
                <CustomButton className={'btn-secondary'} onClick={() => handleClick('1')}>Get One Joke!</CustomButton>
                <CustomButton className={'btn-secondary'} onClick={() => handleClick('3')}>Get Three
                    Jokes!</CustomButton>
                <CustomButton className={'btn-secondary'} onClick={() => handleClick('5')}>Get Five Jokes!</CustomButton>
            </FlexDiv>
        </Styles>
    );
};

export default Jokes