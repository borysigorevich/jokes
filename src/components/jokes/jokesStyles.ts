import styled from 'styled-components'

export const Container = styled.div`
  max-width: 1170px;
  padding: 10px 16px;
  margin: 0 auto;
`

export const Select = styled.select`
  margin-bottom: 10px;
  margin-right: 5px;
  padding: 3px;
  border: 1px solid teal;
  border-radius: 4px;
  
  &:focus {
    outline: none;
  }
`

export const List = styled.ul`
  list-style: none;
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 10px;
  padding-left: 0;

  li {
    border: 1px solid teal;
    border-radius: 4px;
    padding: 3px 5px;
  }
`

export const Input = styled.input`
  display: block;
  position: relative;
  left: 50%;
  transform: translate(-50%);
  margin-bottom: 10px;
  width: 230px;
  padding: 3px 6px;
  border: 1px solid teal;
  border-radius: 4px;
  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
  }
  
  &:focus {
    outline: none;
  }
`

export const Button = styled.button`
  position: relative;
  left: 50%;
  transform: translate(-50%);
  padding: 5px 16px;
  border-radius: 5px;
  border: none;
  background-color: teal;
  color: #fff;
  cursor: pointer;
`
