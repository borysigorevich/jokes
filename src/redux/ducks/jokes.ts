import {call, put, takeEvery} from 'redux-saga/effects';
import {API} from '../../api';
import {SagaIterator} from 'redux-saga';

export const SENT_REQUEST = 'axels-test/joke/SENT_REQUEST';

export const REQUEST_FAILURE = 'axels-test/joke/REQUEST_FAILURE';

export const SET_JOKES = 'axels-test/joke/SET_JOKES';
export const GET_JOKES = 'axels-test/joke/GET_JOKES';

export const SET_MORE_JOKES = 'axels-test/joke/SET_MORE_JOKES';
export const GET_MORE_JOKES = 'axels-test/joke/GET_MORE_JOKES';

export const SET_SORTED_JOKES = 'axels-test/joke/SET_SORTED_JOKES';

export const SET_CATEGORIES = 'axels-test/joke/SET_CATEGORIES';
export const GET_CATEGORIES = 'axels-test/joke/GET_CATEGORIES';

export type JokeType = {
	id: number
	joke: string
	categories: Array<string>
}

export type errorType = {
	message: string
}

export type StateType = {
	jokes: Array<JokeType>
	sortedJokes: Array<JokeType>
	categories: Array<string>
	category: string
	initialRender: boolean
	joke: string
	isLoading: boolean
	error: errorType
}

const joke = JSON.parse(localStorage.getItem('joke')!);

export const initialState: StateType = {
	jokes: joke ? [joke] : [],
	sortedJokes: [],
	categories: [],
	category: 'all',
	initialRender: true,
	joke: '',
	isLoading: true,
	error: {message: ''}
};

//ACTIONS TYPES
export type requestFailureType = {
	type: typeof REQUEST_FAILURE,
	payload: string
}

export type sentRequestType = {
	type: typeof SENT_REQUEST
}

export type setJokesType = {
	type: typeof SET_JOKES
	payload: { jokes: Array<JokeType>, category: string }
}

export type setMoreJokesType = {
	type: typeof SET_MORE_JOKES
	payload: Array<JokeType>
}

export type setSortedJokesType = {
	type: typeof SET_SORTED_JOKES
	payload: string
}

export type setCategoriesType = {
	type: typeof SET_CATEGORIES
	payload: Array<string>
}

export type ActionsType =
	setJokesType |
	setMoreJokesType |
	setSortedJokesType |
	setCategoriesType |
	requestFailureType |
	sentRequestType

export type SagaActionsType =
	getJokesType |
	getMoreJokesType |
	getCategoriesType

//REDUCER
const jokeReducer = (state: StateType = initialState, action: ActionsType): StateType => {
	switch (action.type) {
	case SET_JOKES:
		return {
			...state,
			jokes: [...state.jokes, ...action.payload.jokes],
			sortedJokes: [...state.jokes, ...action.payload.jokes].filter(joke => {
				if (action.payload.category === 'all') {
					return joke;
				} else {
					return joke.categories.includes(action.payload.category);
				}
			}),
			initialRender: false,
			isLoading: false
		};
	case SET_CATEGORIES:
		return {
			...state,
			categories: [...state.categories, ...action.payload],
			initialRender: false
		};
	case SET_MORE_JOKES:
		return {
			...state,
			jokes: [...state.jokes, ...action.payload.filter((joke: JokeType) => {
				return !state.jokes.some(stateJoke => {
					return stateJoke.id === joke.id;
				});
			})],
			sortedJokes: [...state.jokes, ...action.payload.filter((joke: JokeType) => {
				return !state.jokes.some(stateJoke => {
					return stateJoke.id === joke.id;
				});
			})].filter(joke => {
				if (state.category === 'all') {
					return joke;
				} else {
					return joke.categories.includes(state.category);
				}
			}),
			isLoading: false,
			error: {message: ''}
		};
	case SET_SORTED_JOKES:
		return {
			...state,
			sortedJokes: state.jokes.filter(joke => {
				if (action.payload === 'all') {
					return joke;
				} else {
					return joke.categories.includes(action.payload);
				}
			}),
			category: action.payload
		};
	case SENT_REQUEST: {
		return {
			...state,
			isLoading: true
		};
	}
	case REQUEST_FAILURE: {

		return {
			...state,
			isLoading: false,
			error: {message: action.payload}
		};
	}
	default:
		return state;
	}
};

//*******ACTIONS CREATORS********

//SENT REQUEST
export const setIsLoading = (): sentRequestType => ({
	type: SENT_REQUEST
});

//REQUEST FAILURE
export const requestFailure = (error: string): requestFailureType => ({
	type: REQUEST_FAILURE,
	payload: error
});

//GET JOKES
export const setJokes = (jokes: Array<JokeType>, category: string): setJokesType => ({
	type: SET_JOKES,
	payload: {jokes, category}
});

export type getJokesType = {
	type: typeof GET_JOKES,
	payload: { category: string, numOfJokes: string }
}

export const getJokes = (category: string, numOfJokes: string): getJokesType => {
	return {
		type: GET_JOKES,
		payload: {category, numOfJokes}
	};
};

//GET MORE JOKES
export const setMoreJokes = (newJokes: Array<JokeType>): setMoreJokesType => ({
	type: SET_MORE_JOKES,
	payload: newJokes
});

export type getMoreJokesType = {
	type: typeof GET_MORE_JOKES
	payload: {
		numOfJokes: string
		category: string
	}
}

export const getMoreJokes = (numOfJokes: string, category: string): getMoreJokesType => ({
	type: GET_MORE_JOKES,
	payload: {numOfJokes, category}
});

//SORT JOKES
export const setSortedJokes = (category: string): setSortedJokesType => ({
	type: SET_SORTED_JOKES,
	payload: category
});

//GET CATEGORIES
export const setCategories = (categories: Array<string>): setCategoriesType => ({
	type: SET_CATEGORIES,
	payload: categories
});

type getCategoriesType = {
	type: typeof GET_CATEGORIES
}

export const getCategories = (): getCategoriesType => {
	return {
		type: GET_CATEGORIES
	};
};
//*********************

//SAGA WORKERS
export function* getJokesWorker(action: getJokesType): SagaIterator {
	try {
		const res: Array<JokeType> = yield call(API.getJokes, action.payload.numOfJokes, action.payload.category);
		yield put(setJokes(res, action.payload.category));
	} catch (e) {
		yield put(requestFailure((e as Error).message));
	}
}

export function* getCategoriesWorker(): SagaIterator {
	try {
		const res: Array<string> = yield call(API.getCategories);
		yield put(setCategories(res));
	} catch (e) {
		yield put(requestFailure((e as Error).message));
	}
}

export function* getMoreJokesWorker(action: getMoreJokesType): SagaIterator {
	try {
		const res: Array<JokeType> = yield call(API.getMoreJokes, action.payload.numOfJokes, action.payload.category);
		yield put(setMoreJokes(res));
	} catch (e) {
		yield put(requestFailure((e as Error).message));
	}
}

//SAGA WATCHERS
export function* jokesWatcher() {
	yield takeEvery(GET_JOKES, getJokesWorker);
	yield takeEvery(GET_CATEGORIES, getCategoriesWorker);
	yield takeEvery(GET_MORE_JOKES, getMoreJokesWorker);
}

export default jokeReducer;
