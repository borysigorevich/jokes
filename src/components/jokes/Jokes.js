import React, {useEffect, useRef, useState} from 'react'
// import {Button, Col, Container, ListGroup, Row} from "react-bootstrap";
import axios from "axios";
import {Link} from "react-router-dom";
import {Button, Container, Input, List, Select} from "./jokesStyles";


const Jokes = () => {
    const [jokes, setJokes] = useState([])
    const [sortedJokes, setSortedJokes] = useState([])

    const [numOfJokes, setNumOfJokes] = useState('')

    const [categories, setCategories] = useState([])
    const [category, setCategory] = useState('all')

    const initialRender = useRef(false)

    useEffect(() => {
        const getJokes = async () => {
            if (!initialRender.current) {
                try {
                    const res = await axios.get('http://api.icndb.com/jokes/random/3')
                    initialRender.current = true
                    setJokes(res.data.value)
                    setSortedJokes(res.data.value)
                } catch (err) {
                    console.log(err)
                }
            } else {
                // console.log('else')
            }
        }
        getJokes()
    }, [])

    useEffect(() => {
        const getCategories = async () => {
            const res = await axios.get('http://api.icndb.com/categories')
            setCategories(res.data.value)
        }
        getCategories()
    }, [])

    const handleClick = async () => {
        const res = await axios.get(`http://api.icndb.com/jokes/random/${numOfJokes ? numOfJokes : 3}`)
        const receivedJokes = res.data.value

        setJokes([...jokes, ...receivedJokes])
        setSortedJokes([...jokes, ...receivedJokes].filter(joke => {
            if (category === 'all') {
                return joke
            } else {
                return joke.categories.includes(category)
            }
        }))
    }

    const handleChange = (e) => {
        setNumOfJokes(e.target.value)
    }

    const onHandleCategoriesChange = e => {
        const category = e.target.value
        setCategory(category)
        if (category === 'all') {
            setSortedJokes(jokes)
        } else {
            setSortedJokes(jokes.filter(joke => {
                return joke.categories.includes(category)
            }))
        }
    }

    return (
        <Container>
            <Select onChange={onHandleCategoriesChange}>
                <option value="all">All</option>
                {categories.map(category => {
                    return <option key={category}
                                   value={category}>{category[0].toUpperCase() + category.substring(1)}</option>
                })}
            </Select>
            <List>
                {sortedJokes.map(joke => (
                    <li key={joke.id}>
                        <Link className={'link'} to={`/joke/${joke.id}`}>
                            {joke.joke.length > 100 ? joke.joke.substring(0, 100) + '...' : joke.joke}
                        </Link>
                    </li>))
                }
            </List>
            <Input onChange={handleChange}
                   value={numOfJokes}
                   type="number"
                   placeholder={'How many jokes do you want to get?'}/>
            <Button onClick={handleClick}>Get more jokes!</Button>
        </Container>
    );
};

export default Jokes