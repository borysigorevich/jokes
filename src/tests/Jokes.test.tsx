import Enzyme, {shallow} from 'enzyme';
import EnzymeAdapter from '@wojtekmaj/enzyme-adapter-react-17';
import {Jokes} from '../components/jokes/Jokes';
import {
	GET_MORE_JOKES,
	SagaActionsType, SENT_REQUEST,
	sentRequestType,
	setSortedJokesType,
	StateType
} from '../redux/ducks/jokes';
import {mockStore} from './utils';

Enzyme.configure({adapter: new EnzymeAdapter()});

export type actionsType = SagaActionsType | setSortedJokesType | sentRequestType

export type PropsType = {
    jokesState: StateType
    dispatch: (action: actionsType) => void
}

const createJokesComponent = (props: PropsType) => shallow(<Jokes {...props}/>);

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useParams: () => ({
		joke: '2',
		category: 'all'
	}),
	useNavigate: () => ({
		push: jest.fn()
	})
}));

describe('Jokes', () => {

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('snapshot', () => {
		const component = createJokesComponent(mockStore({}));
		expect(component).toMatchSnapshot();
	});

	it('should be one joke', () => {
		const component = createJokesComponent(mockStore({sortedJokes: [{id: 1, joke: 'haha', categories: ['nerdy']}]}));
		expect(component.find('.link').length).toEqual(1);
	});

	it('should be 2 categories', () => {
		const component = createJokesComponent(mockStore({categories: ['nerdy', 'explicit']}));
		expect(component.find('Styled(NavLink)').length).toEqual(3);
	});

	it('should be 2 calls of the dispatch on get joke button click', () => {
		const dispatched: Array<actionsType> = [];
		const component = createJokesComponent(mockStore({dispatch: (action: actionsType) => dispatched.push(action)}));
		const btn = component.find('Styled(Button)').at(0);
		btn.simulate('click');
		expect(dispatched).toEqual([{type: SENT_REQUEST}, {type: GET_MORE_JOKES, payload: '1'}]);
	});
});