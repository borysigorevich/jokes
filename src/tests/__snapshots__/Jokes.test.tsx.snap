// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jokes snapshot 1`] = `
<styled.div>
  <Styled(Navbar)
    className="justify-content-between"
    expand="md"
  >
    <Styled(NavbarBrand)>
      Choose the category
    </Styled(NavbarBrand)>
    <NavbarToggle
      aria-controls="navbar"
      label="Toggle navigation"
    />
    <NavbarCollapse
      className="justify-content-end"
      id="navbar"
    >
      <Nav
        fill={false}
        justify={false}
      >
        <Styled(NavLink)
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "render": [Function],
            }
          }
          className="mb-2 mt-3 mt-sm-0 mb-md-0 me-md-2"
          key="all"
          to="/all"
        >
          All
        </Styled(NavLink)>
      </Nav>
    </NavbarCollapse>
  </Styled(Navbar)>
  <CardGroup
    className="justify-content-center"
  >
    <ListGroup
      as="ul"
      className="w-100"
    >
      <p
        className="fs-3 text-secondary"
      >
        Ooops!! There are no jokes! Get Some!
      </p>
    </ListGroup>
    <Modal
      animation={true}
      autoFocus={true}
      backdrop="static"
      dialogAs={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      enforceFocus={true}
      keyboard={false}
      onHide={[Function]}
      restoreFocus={true}
      show={true}
    >
      <ModalHeader
        closeButton={true}
        closeLabel="Close"
      >
        <ModalTitle>
          Joke
        </ModalTitle>
      </ModalHeader>
      <ModalBody />
      <ModalFooter>
        <Button
          active={false}
          disabled={false}
          onClick={[Function]}
          variant="secondary"
        >
          Close
        </Button>
      </ModalFooter>
    </Modal>
  </CardGroup>
  <styled.div>
    <Spinner
      animation="border"
    />
  </styled.div>
  <styled.div>
    <Styled(Button)
      className="btn-secondary"
      onClick={[Function]}
    >
      Get One Joke!
    </Styled(Button)>
    <Styled(Button)
      className="btn-secondary"
      onClick={[Function]}
    >
      Get Three Jokes!
    </Styled(Button)>
    <Styled(Button)
      className="btn-secondary"
      onClick={[Function]}
    >
      Get Five Jokes!
    </Styled(Button)>
  </styled.div>
</styled.div>
`;
